### PR TITLE
Optimize flatten and descendent traversal

### LINKED ISSUE #2071: Slow performance on large Questionnaire
**Describe the bug**
When the SDC library tries to render a large Questionnaire (+75K lines of JSON). 
1. It takes a long time to load and show the initial page the user can see. To the point of the app becoming not responding, sometimes
2. ~2 second delays for simple action (next button, radios, checkboxes, tapping a keyboard)
3. 3-5 second delay when loading a page w heavy calculation or tapping Start IPC option (which will open up many pages in the background, the next pages)

**Component**
SDC library

**To Reproduce**
1. Download the sample Questionnaire I've provided
3. Paste the downloaded sample Questionnaire into catalog app's `assets` folder
4. Change this line to 'ipc-session-1.json' https://github.com/google/android-fhir/blob/master/catalog/src/main/java/com/google/android/fhir/catalog/LayoutListViewModel.kt#L41
5. Run catalog app
6. Tap 'Layouts' at the bottom app bar
7. Tap 'Default' on the screen
8. Now wait
9. If the app shows not responding, tap wait
10. Once it's rendered, try to interact with the questionnaire, it should have the delays as I mentioned

**Expected behavior**
1. Load the Questionnaire faster, and not to the point of the device went not responding
2. No delays when performing simple action (next button, radios, checkboxes, tapping a keyboard)
3. Less delays for heavy calculation or tapping Start IPC option (which will open up many pages in the background, the next pages)

**Videos**
Demo of the +75K Questionnaire:

[device-2023-07-12-155717.webm](https://github.com/google/android-fhir/assets/62053304/f841e195-95a7-4a60-8093-68fd48c2c889)

Demo of normal sized Questionnaire (In-comparison):

[device-2023-07-12-160004.webm](https://github.com/google/android-fhir/assets/62053304/f3cefb24-f9d1-4898-a6de-bb1d14089497)

**Smartphone (please complete the following information):**
 - Device: Pixel 4 emulator with 3GB of RAM
 - Android version: [e.g. Settings -> About phone -> Android version]
 - Build number: [e.g. Settings -> About phone -> Build number]

**Additional context**
- The sample Questionnaire https://drive.google.com/file/d/1IJsrTvZ5StjJJM_UinsQcQE91j_qE-AZ/view?usp=sharing

**Would you like to work on the issue?**
N/A

### PR DESCRIPTION
**Pull request by omarismail94**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #2071

**Description**
Append to a mutable list, rather than creating a new list at every recursive call and then adding all its elements to the parent list

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: (Bug fix | Feature | Documentation | Testing | **Code health** | Builds | Releases | Other)

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaireItemComponents.kt [hunk 1]
```diff
@@ -670,7 +670,18 @@ internal fun Questionnaire.QuestionnaireItemComponent.extractAnswerOptions(
  */
 fun List<Questionnaire.QuestionnaireItemComponent>.flattened():
   List<Questionnaire.QuestionnaireItemComponent> {
-  return this + this.flatMap { it.item.flattened() }
+  val flattenedList = mutableListOf<Questionnaire.QuestionnaireItemComponent>()
+  this.flattenInto(flattenedList)
+  return flattenedList
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaireItemComponents.kt:675]: feels like a good place to use `also`?

File: datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaireResponseItemComponents.kt [hunk 2]
```diff
@@ -25,11 +25,19 @@ import org.hl7.fhir.r4.model.QuestionnaireResponse
 val QuestionnaireResponse.QuestionnaireResponseItemComponent.descendant:
   List<QuestionnaireResponse.QuestionnaireResponseItemComponent>
   get() {
-    return listOf(this) +
-      this.item.flatMap { it.descendant } +
-      this.answer.flatMap { answer -> answer.item.flatMap { it.descendant } }
+    val preOrderedList = mutableListOf<QuestionnaireResponse.QuestionnaireResponseItemComponent>()
+    appendDescendantTo(preOrderedList)
+    return preOrderedList
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaireResponseItemComponents.kt:30]: same here?

**Review by jingtang10 — APPROVED** [review]

**Comment by jingtang10** [comment]
update the pr description to code health

File: datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaireItemComponents.kt [hunk 3]
```diff
@@ -670,7 +670,18 @@ internal fun Questionnaire.QuestionnaireItemComponent.extractAnswerOptions(
...
+}
+
+private fun List<Questionnaire.QuestionnaireItemComponent>.flattenInto(
+  output: MutableList<Questionnaire.QuestionnaireItemComponent>
+) {
+  this.forEach {
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaireItemComponents.kt:681]: `this` is redundant?

File: datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaireResponseItemComponents.kt [hunk 4]
```diff
@@ -25,11 +25,19 @@ import org.hl7.fhir.r4.model.QuestionnaireResponse
...
   }
 
+private fun QuestionnaireResponse.QuestionnaireResponseItemComponent.appendDescendantTo(
+  output: MutableList<QuestionnaireResponse.QuestionnaireResponseItemComponent>
+) {
+  output.add(this)
+  this.item.forEach { it.appendDescendantTo(output) }
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaireResponseItemComponents.kt:37]: and `this` and below.

**Review by jingtang10 — APPROVED** [review]

**Comment by omarismail94** [comment]
Thanks @jingtang10 ! Updated description and applied your changes

**Pull request merged**